Fix for TouchableNativeFeedback having Animated.Component direct child

tonysherbondy opened the pull request:
Fixes #7996.
Test included.
Not sure this is the best way to go, just a simple solution since the TouchableNativeFeedback is trying to clone the component with a Native component, then seems like it should wrap it with Animated.Component if the incoming child was.

facebook-github-bot-0 commented:
By analyzing the blame information on this pull request, we identified @spicyj and @Kureev to be potential reviewers.

facebook-github-bot-0 commented:
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

facebook-github-bot-3 commented:
@tonysherbondy updated the pull request - [view changes](https://github.com/facebook/react-native/pull/10081/files/263cd9d5366a805af5f02b5c77b35ad286df90b7..b3ee3f1043ea547db7d27401f856a3a2eab7778e)

facebook-github-bot-6 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

tonysherbondy commented:
ping

ide commented:
@vjeux is perhaps the best person to comment on this. My sense is that we don't want to Animated and Touchable, the Animated view should be a child of the Touchable component.

tonysherbondy commented:
Thanks for the feedback @ide. 
My main reasons for making the returned view merge both the Touchable and Animated apis was that 1) the TouchableNative is basically removing the onlyChild in replace of a Touchable native view so it seemed that if the onlyChild was animated then it too should be animated and 2) also seemed like this followed the logic of exposing Animated api for [TouchableOpacity](https://github.com/facebook/react-native/blob/master/Libraries/Components/Touchable/TouchableOpacity.js#L162)

tonysherbondy commented:
ping

tonysherbondy commented:
Ping @vjeux? I have a few days this week available if I can help move this along in some way.

vjeux commented:
@janicduplessis what do you think about this solution? It seems bad design to have this component know about Animated. Would you mind reviewing this pull request? Thanks!

vjeux commented:
@tonysherbondy Thanks for sending this pull request and sorry it's taking so long to get it reviewed :(

janicduplessis commented:
I agree that this component doesn't have to know about Animated. I think this is more working around the issue instead of fixing the cause.

The problem is that props of this component's child gets frozen so there's an error because animated values mutate themselves.

I think ideally we'd want to not to freeze props here. Not sure where this happens and if it is actually needed. Any idea @vjeux?

janicduplessis commented:
Ok, I looked at this more and I understand why this happens. The problem is that we're passing animated props to a non-animated view.

vjeux commented:
Why is this happening? Is it a programming error of the developer building the React Native app, or are we doing something clowny inside of React Native?

janicduplessis commented:
Pretty sure we're doing something clowny :)

janicduplessis commented:
If you look at https://github.com/tonysherbondy/react-native/blob/b3ee3f1043ea547db7d27401f856a3a2eab7778e/Libraries/Components/Touchable/TouchableNativeFeedback.android.js#L218, we are passing all the props from the child view to a native View component. So if the child is not actually a View, in our case an Animated.View, it breaks.

janicduplessis commented:
I suggest we just wrap the child in a new View like we do in other touchable components instead of trying to monkey patch the child. We could also try using React.cloneElement instead but that may not be idea because we add a bunch of props that may already be defined on the child view and overring them.

vjeux commented:
Agreed for adding a wrapper view. Do you want to send a PR for it?

janicduplessis commented:
@tonysherbondy Can you update this PR with the changes we discussed? You can check out the other Touchable components and do something similar.

tonysherbondy commented:
@janicduplessis @vjeux happy to take a pass at this. Thanks for the review!

tonysherbondy commented:
@janicduplessis There is a difficulty with your proposed solution to wrap the children of the TouchableNativeFeedback as we do with TouchableHighlight and others.

The ripple effect is created on Android by changing the background of the TouchableNative view we create with the ripple drawable (https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewManager.java#L104). Thus, any children that we have within the inserted TouchableView will occlude the effect as they are rendered on top.  Touchable highlight has the same problem and gets around this by altering the opacity of the children as it is touched. We could of course do that for TouchableNative, but I think that defeats the point as it would certainly not look like the native feedback when we do that.

janicduplessis commented:
@tonysherbondy Would it work if we use React.cloneElement to add the background to the view?

tonysherbondy commented:
So something like 
```
React.cloneElement(
  child,
  child.props,
  <TouchableView>{child.props.children}</TouchableView>
)
```

janicduplessis commented:
I was thinking just
```
React.cloneElement(
  child,
  {
    backgroundProp: ...
    otherTouchableProps ...
  }
);
```

tonysherbondy commented:
@janicduplessis This works for the desired behavior well. Not sure about what prop clobbering might need to be tested in addition now that we are cloning the original element and overriding some of those props. I've rebased with master and cleaned up the rest of the code since we aren't creating a new view.

tonysherbondy commented:
@janicduplessis if there is something I can do to move this along this week, please let me know.

ide commented:
LGTM

@facebook-github-bot shipit
